### DISCUSSION TITLE
Mapping of POIs

### DISCUSSION THREAD

**Discussion by rinigus** [onthegomap/planetiler#433] [Q&A] [answered]
I am revising MapLibre GL styles that are used in my projects and it is not very clear to me how `class` and `subclass` of POIs is assigned. In particular, I cannot figure out on the basis of Planetiler code and its submodule (https://github.com/openmaptiles/planetiler-openmaptiles.git) why `shop=hairdresser` leads to class and subclass both equal to `hairdresser`. On the basis of the docs (https://openmaptiles.org/schema/#poi), I would have expected class=shop and subclass=hairdresser. When comparing with the `clothing_store` or `alcohol_shop`, there is no special handling either.

Example case:

- OSM: https://www.openstreetmap.org/node/7416834049
- Planetiler import: POI at the center of https://onthegomap.github.io/planetiler-demo/#22/59.435946/24.7466066

Similar seems to happen on data source as used in maptiler.com (not sure if they use Planetiler or older import scripts).

**Comment by msbarry** [comment, marked as answer]
Those come from processing openmaptiles yaml files, for example this file tells openmaptiles and planetiler-openmaptiles which elements to choose, and how to map the tags from those elements to columns in intermediate sql tables:

https://github.com/openmaptiles/openmaptiles/blob/master/layers/poi/mapping.yaml

For `shop=hairdresser`, `hairdresser` is a value it extracts for `shop` attribute, then these lines map the key that produced the match (`shop`) to the `mapping_key` column and value (`hairdresser`) to the `subclass` column which gets passed through to the output:

https://github.com/openmaptiles/openmaptiles/blob/cd77b07e469b8f7726c286723e81d4b24ff7d5cc/layers/poi/mapping.yaml#L341-L344

```yaml
  - name: subclass
    type: mapping_value
  - name: mapping_key
    type: mapping_key
```

Then this file groups certain values from the `subclass` to produce values to put in the `class` attribute:

https://github.com/openmaptiles/openmaptiles/blob/cd77b07e469b8f7726c286723e81d4b24ff7d5cc/layers/poi/poi.yaml#L18-L28

```yaml
      values:
        shop:
          subclass: ['accessories', 'antiques', 'beauty', 'bed', 'boutique', 'camera', 'carpet', 'charity', 'chemist',
                     'coffee', 'computer', 'convenience', 'copyshop', 'cosmetics', 'garden_centre', 'doityourself',
                     'erotic', 'electronics', 'fabric', 'florist', 'frozen_food', 'furniture', 'video_games', 'video',
                     'general', 'gift', 'hardware', 'hearing_aids', 'hifi', 'ice_cream', 'interior_decoration',
                     'jewelry', 'kiosk', 'locksmith', 'lamps', 'mall', 'massage', 'motorcycle', 'mobile_phone', 
                     'newsagent', 'optician', 'outdoor', 'perfumery', 'perfume', 'pet', 'photo', 'second_hand', 
                     'shoes', 'sports', 'stationery', 'tailor', 'tattoo', 'ticket', 'tobacco', 'toys', 'travel_agency', 
                     'watches', 'weapons', 'wholesale']
        town_hall:
```

The planetiler-openmaptiles repo has a script that compiles those yaml scripts to java code:

https://github.com/openmaptiles/planetiler-openmaptiles/blob/b39e0b30e1de0fd672598bcc9e7aa89294e85a9c/src/main/java/org/openmaptiles/generated/Tables.java#L708-L750

```java
    public static final Expression MAPPING = and(or(...
      matchAny("shop", "accessories", "alcohol", "antiques", "art", "bag", "bakery", "beauty", "bed", "beverages",
        "bicycle", "books", "boutique", "butcher", "camera", "car", "car_repair", "car_parts", "carpet", "charity",
        "chemist", "chocolate", "clothes", "coffee", "computer", "confectionery", "convenience", "copyshop",
        "cosmetics", "deli", "delicatessen", "department_store", "doityourself", "dry_cleaning", "electronics",
        "erotic", "fabric", "florist", "frozen_food", "furniture", "garden_centre", "general", "gift", "greengrocer",
        "hairdresser", "hardware", "hearing_aids", "hifi", "ice_cream", "interior_decoration", "jewelry", "kiosk",
        "lamps", "laundry", "locksmith", "mall", "massage", "mobile_phone", "motorcycle", "music", "musical_instrument",
        "newsagent", "optician", "outdoor", "perfume", "perfumery", "pet", "photo", "second_hand", "shoes", "sports",
        "stationery", "supermarket", "tailor", "tattoo", "ticket", "tobacco", "toys", "travel_agency", "video",
        "video_games", "watches", "weapons", "wholesale", "wine")), matchType("point"));
```

https://github.com/openmaptiles/planetiler-openmaptiles/blob/b39e0b30e1de0fd672598bcc9e7aa89294e85a9c/src/main/java/org/openmaptiles/generated/OpenMapTilesSchema.java#L1793-L1843

```java
    final class FieldMappings {
      public static final MultiExpression<String> Class = MultiExpression.of(List.of(
        MultiExpression.entry("shop",
          matchAny("subclass", "accessories", "antiques", "beauty", "bed", "boutique", "camera", "carpet", "charity",
            "chemist", "coffee", "computer", "convenience", "copyshop", "cosmetics", "garden_centre", "doityourself",
            "erotic", "electronics", "fabric", "florist", "frozen_food", "furniture", "video_games", "video", "general",
            "gift", "hardware", "hearing_aids", "hifi", "ice_cream", "interior_decoration", "jewelry", "kiosk",
            "locksmith", "lamps", "mall", "massage", "motorcycle", "mobile_phone", "newsagent", "optician", "outdoor",
            "perfumery", "perfume", "pet", "photo", "second_hand", "shoes", "sports", "stationery", "tailor", "tattoo",
            "ticket", "tobacco", "toys", "travel_agency", "watches", "weapons", "wholesale")),
        MultiExpression.entry("town_hall",
          matchAny("subclass", "townhall", "public_building", "courthouse", "community_centre")), ...);
    }
```

If you think you found an issue, I'd suggest fixing it in the openmaptiles repo yaml files, then on the next version upgrade planetiler-openmaptiles will get that fix automatically.  Or you could add logic that falls back to setting the class to `mapping_key` so it will be `shop` for any value of `shop=...` that matches

**Reply by rinigus** [reply]
Thank you very much for a fast and detailed reply. So, in this case, we have `hairdresser` in `mapping.yaml` section that you listed (https://github.com/openmaptiles/openmaptiles/blob/master/layers/poi/mapping.yaml#L176), but not in the later section in poi.yaml (https://github.com/openmaptiles/openmaptiles/blob/cd77b07e469b8f7726c286723e81d4b24ff7d5cc/layers/poi/poi.yaml#L18-L28). As a result, generated Planetiler code is missing `hairdresser` in 
```java
    final class FieldMappings {
      public static final MultiExpression<String> Class = MultiExpression.of(List.of(
        MultiExpression.entry("shop",
          matchAny("subclass", "accessories", "antiques", "beauty", "bed", "boutique", "camera", "carpet", "charity",
            "chemist", "coffee", "computer", "convenience", "copyshop", "cosmetics", "garden_centre", "doityourself",
            "erotic", "electronics", "fabric", "florist", "frozen_food", "furniture", "video_games", "video", "general",
            "gift", "hardware", "hearing_aids", "hifi", "ice_cream", "interior_decoration", "jewelry", "kiosk",
            "locksmith", "lamps", "mall", "massage", "motorcycle", "mobile_phone", "newsagent", "optician", "outdoor",
            "perfumery", "perfume", "pet", "photo", "second_hand", "shoes", "sports", "stationery", "tailor", "tattoo",
            "ticket", "tobacco", "toys", "travel_agency", "watches", "weapons", "wholesale")),
        MultiExpression.entry("town_hall",
          matchAny("subclass", "townhall", "public_building", "courthouse", "community_centre")), ...);
    }
```

Does it mean that if there is no such mapping found, `subclass` of intermediate mapping gets promoted to `class` in Planetiler?

**Reply by msbarry** [reply]
Yep planetiler-openmaptiles mirrors this logic from openmaptiles that falls back to subclass 

https://github.com/openmaptiles/openmaptiles/blob/cd77b07e469b8f7726c286723e81d4b24ff7d5cc/layers/poi/class.sql#L32

Might want to start asking the question in openmaptiles to see if this is a bug or if they might want the fallback to be something besides subclass.

**Reply by rinigus** [reply]
Thank you very much! It was somewhat unexpected, but maybe the logic can be readout from description of `class` in docs. I will think how to formulate what made me confused and open an issue at openmaptiles.
